sec/insecure/insecure.go

```diff
@@ -5,64 +5,172 @@ package insecure
+// LocalPeer returns the transport's local peer ID.
 func (t *Transport) LocalPeer() peer.ID {
 	return t.id
 }
 
-// LocalPrivateKey returns nil. This transport is not secure.
+// LocalPrivateKey returns the local private key.
+// This key is used only for identity generation and provides no security.
 func (t *Transport) LocalPrivateKey() ci.PrivKey {
-	return nil
+	return t.key
 }
 
 // SecureInbound *pretends to secure* an outbound connection to the given peer.
+// It sends the local peer's ID and public key, and receives the same from the remote peer.
+// No validation is performed as to the authenticity or ownership of the provided public key,
+// and the key exchange provides no security.
+//
+// SecureInbound may fail if the remote peer sends an ID and public key that are inconsistent
+// with each other, or if a network error occurs during the ID exchange.
 func (t *Transport) SecureInbound(ctx context.Context, insecure net.Conn) (sec.SecureConn, error) {
-	return &Conn{
-		Conn:  insecure,
-		local: t.id,
-	}, nil
+	conn := &Conn{
+		Conn:         insecure,
+		local:        t.id,
+		localPrivKey: t.key,
+	}
+
+	err := conn.runHandshakeSync(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
 }
 
 // SecureOutbound *pretends to secure* an outbound connection to the given peer.
+// It sends the local peer's ID and public key, and receives the same from the remote peer.
+// No validation is performed as to the authenticity or ownership of the provided public key,
+// and the key exchange provides no security.
+//
+// SecureOutbound may fail if the remote peer sends an ID and public key that are inconsistent
+// with each other, or if the ID sent by the remote peer does not match the one dialed. It may
+// also fail if a network error occurs during the ID exchange.
 func (t *Transport) SecureOutbound(ctx context.Context, insecure net.Conn, p peer.ID) (sec.SecureConn, error) {
-	return &Conn{
-		Conn:   insecure,
-		local:  t.id,
-		remote: p,
-	}, nil
+	conn := &Conn{
+		Conn:         insecure,
+		local:        t.id,
+		localPrivKey: t.key,
+	}
+
+	err := conn.runHandshakeSync(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if p != conn.remote {
+		return nil, fmt.Errorf("remote peer sent unexpected peer ID. expected=%s received=%s",
+			p, conn.remote)
+	}
+
+	return conn, nil
 }
 
 // Conn is the connection type returned by the insecure transport.
 type Conn struct {
 	net.Conn
+
 	local  peer.ID
 	remote peer.ID
+
+	localPrivKey ci.PrivKey
+	remotePubKey ci.PubKey
+}
+
+func makeExchangeMessage(privkey ci.PrivKey) (*pb.Exchange, error) {
+	pubkey, err := ci.PublicKeyToProto(privkey.GetPublic())
+	if err != nil {
+		return nil, err
+	}
+	id, err := peer.IDFromPrivateKey(privkey)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.Exchange{
+		Id:     []byte(id),
+		Pubkey: pubkey,
+	}, nil
+}
+
+func (ic *Conn) runHandshakeSync(ctx context.Context) error {
+	const maxSize = 1 << 16
```

Comment:
we have a default here https://github.com/libp2p/go-libp2p-core/blob/master/network/network.go#L23

but honestly we can probably place a smaller limit on key exchanges. just linking for info

Reply:
on second thought, may be worth just using that constant. also this is reminding me, one of msgio's big benefits is buffer pooling, but since this is a test-only upgrader, we're probably fine with a little extra allocation.

Reply:
awesome, thanks for pointing me at the constant. it seems best to use it for consistency rather than have more magic numbers sprinkled about.